src/memory-analyzer/gdb_api.cpp

```diff
@@ -0,0 +1,491 @@
+/*******************************************************************\
+
+Module: GDB Machine Interface API
+
+Author: Malte Mues <mail.mues@gmail.com>
+        Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Low-level interface to gdb
+
+// clang-format off
+#if defined(__linux__) || \
+    defined(__FreeBSD_kernel__) || \
+    defined(__GNU__) || \
+    defined(__unix__) || \
+    defined(__CYGWIN__) || \
+    defined(__MACH__)
+// clang-format on
+
+#include <cctype>
+#include <cerrno>
+#include <cstdio>
+#include <cstring>
+#include <regex>
+
+#include "gdb_api.h"
+
+#include <goto-programs/goto_model.h>
+
+#include <util/string_utils.h>
+
+/// Create a gdb_apit object
+///
+/// \param binary the binary to run with gdb
+/// \param log boolean indicating whether gdb input and output should be logged
+gdb_apit::gdb_apit(const char *binary, const bool log)
+  : binary(binary), log(log), gdb_state(gdb_statet::NOT_CREATED)
+{
+}
```

Comment:
I spoke face-to-face with @danpoe to ask whether a better model was to have create_gdb_process be private, and have the constructor be responsible for starting the process. I honestly don't know which is the best approach, but I'm leaving a comment here for the record.